Improve Tuya humidifier error handling and diagnostics for unsupporte…

alunduil opened the pull request:
…d devices (#119865)

- Skip entity creation for humidifier devices missing required DPCodes
- Add validation in _send_command to prevent sending empty code commands
- Enhance error messages with detailed device diagnostics information
- Add comprehensive logging to help identify device compatibility patterns
- Add tests covering devices without required control codes

This improves diagnostic information for issue #119865 to help determine if the problem is systematic or varies by device model, enabling better future support improvements.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 119865
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

alunduil committed:
Improve Tuya humidifier error handling and diagnostics for unsupported devices (#119865)

- Skip entity creation for humidifier devices missing required DPCodes
- Add validation in _send_command to prevent sending empty code commands
- Enhance error messages with detailed device diagnostics information
- Add comprehensive logging to help identify device compatibility patterns
- Add tests covering devices without required control codes

This improves diagnostic information for issue #119865 to help determine
if the problem is systematic or varies by device model, enabling better
future support improvements.

joostlek commented on homeassistant/components/tuya/entity.py:
```diff
@@ -177,5 +177,19 @@ async def _handle_state_update(
 
     def _send_command(self, commands: list[dict[str, Any]]) -> None:
         """Send command to the device."""
+        # Validate commands before sending
+        for command in commands:
+            if command.get("code") is None:
+                raise ValueError(
+                    f"Cannot send command with empty 'code' to device {self.device.id} "
+                    f"({self.device.name}). Command: {command}. This indicates the device "
+                    f"does not support the requested operation or Home Assistant cannot "
+                    f"find the correct control method. Available device capabilities: "
+                    f"function codes: {list(self.device.function.keys())}, status codes: "
+                    f"{list(self.device.status.keys())}. Your device may have limited "
+                    f"remote control support compared to what's available in the Smart Life "
+                    f"or Tuya Smart apps."
+                )
```
When does this happen? Shouldn't we rather prevent these commands to be sent in the first place?

alunduil replied:
Thanks for the reminder, this had moved into other verifications and I forgot to revert this. I'll remove it now.

epenet commented on tests/components/tuya/fixtures/cs_broken_dehumidifier_missing_switch.json:
```diff
@@ -0,0 +1,37 @@
+  "id": "bf3fce6af592f12df3broken",
+  "name": "Broken Dehumidifier",
+  "category": "cs",
+  "product_id": "broken123",
```
It is really import to use "real" diagnostic dumps, and I have reviewed https://github.com/home-assistant/core/issues/119865 but I couldn't see there a diagnostic dump that matched this fixture.

I have in parallel opened https://github.com/home-assistant/core/pull/148797 to add two new fixtures for the new product ids.

alunduil replied:
Thanks for that, I'm still learning what best practices are here. I'll grab the device_id from my personal unit that suffers from this problem and update this fixture accordingly. Please let me know if you see anything else that is out of place so I can resolve it.

alunduil replied:
Let me know if I ended up where you would expect. If not, I'll make one more pass and ensure it's correct.

alunduil commented:
@joostlek and @epenet thanks for the comments, I'll review them shortly and integrate those good suggestions. I'll mark this as ready for review again when I've finished that.

alunduil commented on tests/components/tuya/__init__.py:
```diff
@@ -26,6 +26,9 @@
         Platform.SWITCH,
     ],
+    "cs_broken_dehumidifier_missing_switch": [
+        # No platforms supported - missing required switch DPCodes
```
I'm not sure what that entails so I might need a bit more guidance to make that happen but I'm happy to include that in this pull request.

alunduil committed:
fixup tests for humidifiers

This should be squashed after confirmation that these fixes are acceptable. I'm
keeping them separate to make reviewing the test updates simpler and on their
own for further fixes and reviews.

epenet commented:
> I intend to squash if they look good

Please don't... it makes reviewing harder.
The squash will be done automatically on merge

epenet commented:
I think you need to compile the translations before updating the snapshots

alunduil commented on homeassistant/components/tuya/humidifier.py:
```diff
@@ -108,6 +119,28 @@ def __init__(
+                "No suitable switch DPCode found for %s device '%s' (device_id: %s, category: %s). "
+                "Expected one of: %s. Available DPCodes: %s. "
+                "This device may not be controllable through Home Assistant. "
+                "Note that some devices have more features available in the Smart Life or Tuya Smart "
+                "apps than can be controlled remotely through Home Assistant. Consider reporting this "
+                "issue with your device diagnostics to help improve support",
```
That sounds good. I'll drop that and leave it at the notice about not being controllable.

alunduil committed:
fixup comments from pull request #119865

epenet commented on homeassistant/components/tuya/humidifier.py:
```diff
@@ -169,10 +202,24 @@ def current_humidity(self) -> int | None:
     def turn_on(self, **kwargs: Any) -> None:
         """Turn the device on."""
+        if not self.has_required_dpcodes:
+            raise ServiceValidationError(
```
IMO all these raise changes should be in a separate PR, and aligned for all platforms. 
Most still raise RuntimeError and I think it should be changed for all

alunduil commented on homeassistant/components/tuya/humidifier.py:
```diff
@@ -71,9 +72,20 @@ def async_discover_device(device_ids: list[str]) -> None:
-                    TuyaHumidifierEntity(device, hass_data.manager, description)
-                )
+                # Check if device has any of the required DPCodes before creating entity
+                entity = TuyaHumidifierEntity(device, hass_data.manager, description)
```
It looks like if I don't create the entity I have to duplicate the logic of find_dpcode here which smells wrong to me. I can do that, but what do you suggest as a better pattern? The difficulty I see comes from the union type over DPCode and tuples of DPCodes. Happy to make that change but it doesn't seem as straightforward and readable as just making the entity (which will be garbage collected if we don't end up using it anyway).

Thanks in advance for any clarification here.

epenet replied:
Since `description.dpcode` is always set, the easiest would be to check something like this:
`if description := HUMIDIFIERS.get(device.category) and description.dpcode is not None and any(code in device.status for code in description.dpcode)`
